bee-message/src/message.rs

```diff
@@ -0,0 +1,217 @@
+
+use crypto::hashes::{blake2b::Blake2b256, Digest};
+
+/// The minimum number of bytes in a message.
+pub const MESSAGE_LENGTH_MIN: usize = 53;
+
+/// The maximum number of bytes in a message.
+pub const MESSAGE_LENGTH_MAX: usize = 32768;
+
+/// Represent the object that nodes gossip around the network.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Message {
+    /// Specifies which network this message is meant for.
+    network_id: u64,
+    /// The [MessageId]s that this message directly approves.
+    parents: Parents,
+    /// The optional [Payload] of the message.
+    payload: Option<Payload>,
+    /// The result of the Proof of Work in order fot the message to be accepted into the tangle.
+    nonce: u64,
+}
+
+impl Message {
+    /// Create a new `MessageBuilder` to construct an instance of a `Message`.
+    pub fn builder() -> MessageBuilder {
+        MessageBuilder::new()
+    }
+
+    /// Computes the identifier of the message.
+    pub fn id(&self) -> (MessageId, Vec<u8>) {
+        let bytes = self.pack_new();
+        let id = Blake2b256::digest(&bytes);
+
+        (MessageId::new(id.into()), bytes)
+    }
+
+    /// Return the network id of a `Message`.
+    pub fn network_id(&self) -> u64 {
+        self.network_id
+    }
+
+    /// Return the parents of a `Message`.
+    pub fn parents(&self) -> &Parents {
+        &self.parents
+    }
+
+    /// Return the optional payload of a `Message`.
+    pub fn payload(&self) -> &Option<Payload> {
+        &self.payload
+    }
+
+    /// Return the nonce of a `Message`.
+    pub fn nonce(&self) -> u64 {
+        self.nonce
+    }
+}
+
+impl Packable for Message {
+    type Error = Error;
+
+    fn packed_len(&self) -> usize {
+        self.network_id.packed_len()
+            + self.parents.packed_len()
+            + option_payload_packed_len(self.payload.as_ref())
+            + self.nonce.packed_len()
+    }
+
+    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
+        self.network_id.pack(writer)?;
+        self.parents.pack(writer)?;
+        option_payload_pack(writer, self.payload.as_ref())?;
+        self.nonce.pack(writer)?;
+
+        Ok(())
+    }
+
+    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error> {
+        let network_id = u64::unpack_inner::<R, CHECK>(reader)?;
+
+        let parents = Parents::unpack_inner::<R, CHECK>(reader)?;
+
+        let (payload_len, payload) = option_payload_unpack::<R, CHECK>(reader)?;
+
+        if CHECK
+            && !matches!(
+                payload,
+                None | Some(Payload::Transaction(_)) | Some(Payload::Milestone(_)) | Some(Payload::Indexation(_))
+            )
+        {
+            // Safe to unwrap since it's known not to be None.
+            return Err(Error::InvalidPayloadKind(payload.unwrap().kind()));
```

Comment:
Is `payload` known not to be `None`? It looks like the above statement matches on `payload == None`, unless I'm missing something obvious (which could well be the case!).

Reply:
We check if it `!matches(...)`

Reply:
🤦